flink-runtime/src/main/java/org/apache/flink/runtime/rpc/heartbeat/HeartbeatScheduler.java

```diff
@@ -0,0 +1,286 @@
+	 *
+	 * @return heartbeat interval
+	 */
+	public long getHeartbeatInterval() {
+		return heartbeatInterval;
+	}
+
+	/**
+	 * trigger heartbeat to target gateway
+	 *
+	 * @param leaderID leader session id of current sender
+	 * @param timeout  timeout for heartbeat response
+	 * @return HeartbeatResponsePayload wrapped in future
+	 */
+	protected abstract Future<Payload> triggerHeartbeat(UUID leaderID, FiniteDuration timeout);
+
+	/**
+	 * report heartbeat response payload to sender who sending heartbeat
+	 *
+	 * @param heartbeatResponsePayload heartbeat response which contains payload
+	 */
+	protected abstract void reportHeartbeatPayload(Payload heartbeatResponsePayload);
+
+	/**
+	 * callback method when heartbeat sender lost heartbeat with target
+	 */
+	protected abstract void lossHeartbeat();
+
+	/**
+	 * send a heartbeat attempt to target, receive the response from target or failed depends on the future result.
+	 *
+	 * @param attempt
+	 * @param timeoutMillis
+	 * @param currentHeartbeatBeginTime
+	 */
+	private void sendHeartbeatToTaskManager(final int attempt, final long timeoutMillis, final long currentHeartbeatBeginTime) {
+		// eager check for closed to avoid some unnecessary work
+		if (closed) {
+			return;
+		}
+		FiniteDuration timeout = new FiniteDuration(timeoutMillis, TimeUnit.MILLISECONDS);
+		Future<Payload> heartbeatResponse = triggerHeartbeat(leaderID, timeout);
+
+		heartbeatResponse.onSuccess(new OnSuccess<Payload>() {
+
+			@Override
+			public void onSuccess(Payload result) throws Throwable {
+				if (!isClosed()) {
+					// report heartbeat response payload back to sender
+					reportHeartbeatPayload(result);
+					scheduleNewHeartbeatToTaskManagerLater(heartbeatTimeout, heartbeatInterval);
+				}
+			}
+		}, rpcService.getExecutionContext());
+
+		// upon failure, retry
+		heartbeatResponse.onFailure(new OnFailure() {
+			@Override
+			public void onFailure(Throwable failure) {
+				if (!isClosed()) {
+					long currentTime = System.currentTimeMillis();
+					if(currentTime - currentHeartbeatBeginTime >= maxAttemptTime) {
+						log.error("Lost heartbeat with at {} ({}) after {} attempts", targetName, targetAddress, attempt);
+						closed = true;
+						// mark target as failed after heartbeat interaction attempts failed for max attempt time
+						lossHeartbeat();
+					} else {
+						// we simply have not received a heartbeat response in time. maybe the timeout was
+						// very low (initial fast registration attempts), maybe the target endpoint is
+						// currently down.
+						if (failure instanceof TimeoutException) {
+							if (log.isDebugEnabled()) {
+								log.debug("Heartbeat to {} ({}) attempt {} timed out after {} ms",
+									targetName, targetAddress, attempt, timeoutMillis);
+							}
+
+							long retryTimeoutMillis = Math.min(2 * timeoutMillis, maxHeartbeatTimeout);
+							sendHeartbeatToTaskManager(attempt + 1, retryTimeoutMillis, currentHeartbeatBeginTime);
```

Comment:
till,  long retryTimeoutMillis = Math.min(2 * timeoutMillis, maxHeartbeatTimeout); maybe we could use maxHeartbeatTimeout to avoid this problem? Or we just use fix period instead of exponential backoff?